File: pages/refactorings/replace-mutable-fields-with-lenses/02-player-position/After.scala
// Same, with a composed path: player . x and player . y reach the
// nested field in one step, no copy chain.
object After:
  case class Player(id: String, x: Int, y: Int)
  case class Game(level: Int, player: Player)

  case class Lens[S, A](get: S => A, set: A => S => S):
    def modify(f: A => A): S => S = s => set(f(get(s)))(s)

  def compose[S, M, A](outer: Lens[S, M],
                       inner: Lens[M, A]): Lens[S, A] =
    Lens[S, A](s => inner.get(outer.get(s)),
      a => s => outer.set(inner.set(a)(outer.get(s)))(s))

  val player: Lens[Game, Player] =
    Lens[Game, Player](_.player, p => g => g.copy(player = p))

  val x: Lens[Player, Int] =
    Lens[Player, Int](_.x, v => p => p.copy(x = v))
  val y: Lens[Player, Int] =
    Lens[Player, Int](_.y, v => p => p.copy(y = v))

  val playerX: Lens[Game, Int] = compose(player, x)
  val playerY: Lens[Game, Int] = compose(player, y)

  def moveX(g: Game, dx: Int): Game = playerX.modify(_ + dx)(g)
  def moveY(g: Game, dy: Int): Game = playerY.modify(_ + dy)(g)

File: pages/refactorings/replace-mutable-fields-with-lenses/02-player-position/After.hs
-- Same, with a composed path: player . x and player . y reach the
-- nested field in one step, no copy chain.
module After where

data Player = Player { pid :: String, px :: Int, py :: Int }
  deriving (Eq, Show)
data Game   = Game   { level :: Int, gplayer :: Player }
  deriving (Eq, Show)

data Lens s a = Lens { get :: s -> a, set :: a -> s -> s }

modify :: Lens s a -> (a -> a) -> s -> s
modify l f s = set l (f (get l s)) s

compose :: Lens s m -> Lens m a -> Lens s a
compose outer inner = Lens
  { get = get inner . get outer
  , set = \a s -> set outer (set inner a (get outer s)) s
  }

player :: Lens Game Player
player = Lens { get = gplayer, set = \p g -> g { gplayer = p } }

x :: Lens Player Int
x = Lens { get = px, set = \v p -> p { px = v } }

y :: Lens Player Int
y = Lens { get = py, set = \v p -> p { py = v } }

playerX :: Lens Game Int
playerX = compose player x

playerY :: Lens Game Int
playerY = compose player y

moveX :: Game -> Int -> Game
moveX g dx = modify playerX (+ dx) g

moveY :: Game -> Int -> Game
moveY g dy = modify playerY (+ dy) g
